File: third-semester/homework1.1/LazyEvaluation/SingleThreadedLazy.cs
namespace LazyEvaluation
{
    using System;

    public class SingleThreadedLazy<T> : ILazy<T>
    {
        private readonly Func<T> supplier;
        private bool isEvaluated;
        private T result;

        public SingleThreadedLazy(Func<T> supplier)
        {
            this.isEvaluated = false;
            this.supplier = supplier;
        }

        public T Get()
        {
            if (!this.isEvaluated)
            {
                this.result = this.supplier();
                this.isEvaluated = true;
            }

            return this.result;
        }
    }
}

File: 4th-semester/Homework8.1/Lazy/SingleThreadedLazy.fs
namespace Lazy

/// Single threaded lazy implementation
type SingleThreadedLazy<'a>(supplier: unit -> 'a) =
    /// Evaluation result
    let mutable result = None
    
    interface ILazy<'a> with
        /// Gets evaluation result
        /// Supplier function is called only once
        member this.Get() =
            if result.IsNone then result <- Some (supplier ())
            
            result.Value